core/src/OpenACC/Kokkos_OpenACC_Instance.cpp

```diff
@@ -46,17 +46,40 @@
 
 #include <OpenACC/Kokkos_OpenACC_Instance.hpp>
 #include <OpenACC/Kokkos_OpenACC.hpp>
+#include <OpenACC/Kokkos_OpenACC_Traits.hpp>
 #include <impl/Kokkos_Profiling.hpp>
+#include <impl/Kokkos_DeviceManagement.hpp>
 
-#include <ostream>
+#include <openacc.h>
+
+#include <iostream>
+
+namespace Kokkos {
+bool show_warnings() noexcept;
+}
 
 Kokkos::Experimental::Impl::OpenACCInternal*
 Kokkos::Experimental::Impl::OpenACCInternal::singleton() {
   static OpenACCInternal self;
   return &self;
 }
 
-void Kokkos::Experimental::Impl::OpenACCInternal::initialize() {
+void Kokkos::Experimental::Impl::OpenACCInternal::initialize(
+    InitializationSettings const& settings) {
+  if (OpenACC_Traits::may_fallback_to_host &&
+      acc_get_num_devices(OpenACC_Traits::dev_type) == 0 &&
+      !settings.has_device_id()) {
+    if (show_warnings()) {
+      std::cerr << "Warning: No GPU available for execution, falling back to"
+                   " using the host!"
+                << std::endl;
+    }
+    acc_set_device_type(acc_device_host);
+  } else {
+    using Kokkos::Impl::get_gpu;
+    int const dev_num = get_gpu(settings);
+    acc_set_device_num(dev_num, OpenACC_Traits::dev_type);
+  }
```

Comment:
@seyonglee @masterleinad how does that sound?

Reply:
It sounds reasonable.
By the way, is it possible for `get_gpu(settings)` to return negative value? If so, we may have to insert code to check validity of `dev_num` in line 80 too.

Reply:
Looks good to me. I think we should also check in https://github.com/kokkos/kokkos/blob/0ef409bab190cdae9417bff34f7cf5c90048f227/core/src/impl/Kokkos_Core.cpp#L355-L357 that `settings.get_device_id()` is less than `visible_devices.size()` but that's independent of this pull request.

Reply:
The intent is that `Impl::get_gpu` will do the validation.  See #5253 
As it is currently implemented, you could technically get in trouble when using `KOKKOS_VISIBLE_DEVICES` but as this is an issue for all backends, we will improve this and handle it in "core".